### PR TITLE
Handle invalid connection hash entries

### PR DESCRIPTION
**Pull request by onurctirtir**
DESCRIPTION: Fixes a segfault in connection management due to invalid connection hash entries

Fixes https://github.com/citusdata/citus/issues/4359

----------

If `MemoryContextAlloc` errors out -e.g. during an OOM-, `ConnectionHashEntry->connections`
stays as NULL.

With this commit, we add `isValid` flag to `ConnectionHashEntry` that should be set to `true`
right after we allocate & initialize `ConnectionHashEntry->connections` list properly, and we
check it before accesing to `ConnectionHashEntry->connections`.

----------

See detailed investigation at https://github.com/citusdata/citus/issues/4359

Also, to reproduce/test the issue, readers can use 57e84c7227bd644ed26b24511bfb3542ed56b5e9 (along with the repro steps 
given in https://github.com/citusdata/citus/issues/4359) both on master and on this branch to see what we fix with this pr.

Any ideas to add regression tests ?

### CODE REVIEW

File: src/backend/distributed/connection/connection_management.c [hunk 1]
```diff
@@ -289,11 +295,23 @@ StartNodeUserDatabaseConnection(uint32 flags, const char *hostname, int32 port,
...
 
 	ConnectionHashEntry *entry = hash_search(ConnectionHash, &key, HASH_ENTER, &found);
-	if (!found)
+	if (!found || !entry->isValid)
 	{
+		/*
+		 * We are just building hash entry or previously it was left in an
+		 * invalid state as we couldn't allocate memory for it.
+		 * So initialize entry->connections list here.
+		 */
```
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/connection/connection_management.c:304]: makes a lot sense, done at https://github.com/citusdata/citus/pull/4362/commits/0f2de8a11786dabee48f3084cd15ceb9c454d2fe

File: src/backend/distributed/connection/connection_management.c [hunk 2]
```diff
@@ -161,6 +161,12 @@ AfterXactConnectionHandling(bool isCommit)
 	hash_seq_init(&status, ConnectionHash);
 	while ((entry = (ConnectionHashEntry *) hash_seq_search(&status)) != 0)
 	{
+		if (!entry->isValid)
+		{
+			/* skip invalid connection hash entries */
+			continue;
+		}
+
 		AfterXactHostConnectionHandling(entry, isCommit);
```
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/connection/connection_management.c:170]: >I miss the context when objects are removed from the hash (it seems to never happen)

Yeah, we don't remove them from hash, for example https://github.com/citusdata/citus/blob/6e493624af4eaada536eb1201dd0d82ef53f4b9c/src/backend/distributed/connection/connection_management.c#L166..L170

>How do we normally handle reuse of objects outside of the valid part.

We just continue to use the same doubly-linked list head

[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/connection/connection_management.c:170]: makes sense now, thanks

**Review by thanodnl — APPROVED** [review]
Thanks for addressing this

I have verified we check the valid flag on every read from the connection hash.

**Pull request merged**